src/Statistics.jl

```diff
@@ -165,9 +162,22 @@ julia> mean(A, dims=2)
 ```
 """
 mean(A::AbstractArray; dims=:) = _mean(A, dims)
+# mean(A::AbstractArray; dims=:) = _mean_promote(A, dims=dims)
 
 _mean(A::AbstractArray{T}, region) where {T} = mean!(Base.reducedim_init(t -> t/2, +, A, region), A)
-_mean(A::AbstractArray, ::Colon) = sum(A) / length(A)
+
+_mean(A::AbstractArray, ::Colon) = _mean_promote(A)
+function _mean_promote(A::AbstractArray; map_funct=identity, dims=:)
+    isempty(A) && return sum(map_funct.(A))/0
```

Comment:
```suggestion
    isempty(A) && return sum(map_funct, A)/0
```
Also this doesn't work when `dims !== :`. So better have a separate method for `dims::Colon`. You can define `_prom` in the global scope to use it in all methods (but call it e.g. `_promote_mean`).

Reply:
Wouldn't adding dims=... argument to sum on line 171 fix that for all cases?

Reply:
[As noted by @stevengj](https://github.com/JuliaLang/Statistics.jl/issues/22#issuecomment-587122377), promotion isn't necessary when `dims !== :`. So just keep the existing `_mean` method from above and restrict this one to `dims::Colon`.

Reply:
> [As noted by @stevengj](https://github.com/JuliaLang/Statistics.jl/issues/22#issuecomment-587122377), promotion isn't necessary when `dims !== :`. 

Is that promotion exactly identical in all cases to _mean_promote()?

> Can you add tests for `mean(f, A, dims=...)` methods?

Right now there're two implementations for dims not equal to Colon - line #166 and the one starting at #172. They don't seem to be fully compatible. The new one breaks tests for ```var()```.

Reply:
What I'm driving at, perhaps, the implementation from #166 ought to be eliminated. But that might require fixing errors in var().

Reply:
> Is that promotion exactly identical in all cases to `_mean_promote()`?

Well the situation there is quite messy, and there's a very old comment saying that things should be improved. It calls `Base.reducedim_init`, which fills the array with `z = zero(zero(eltype(A))/2)); z+z` for common numeric types. For other types, it fills the array with `z = zero(zero(_realtype(f, promote_union(eltype(A))))/2); z+z` if `zero` is supported, and otherwise with `zero(sum(t -> T/2, A))`. Overall since `/2` is present in all paths, the promotion behavior should be relatively consistent with `_mean_promote`.

Things will probably differ for heterogeneous types, where `_mean_promote` will progressively promote the type to take into account new entries, while when `dims !== :` promotion will always be done upfront. But we making reduction over dimensions promote progressively would be complex and inefficient (that would imply copying arrays), and in practice cases where results would differ should be quite rare. OTOH we could make the `dims === :` case always promote beforehand using similar rules to `Base.reducedim_init`, but that's probably not worth it.

@stevengj Do you agree?

> Right now there're two implementations for dims not equal to Colon - line #166 and the one starting at #172. They don't seem to be fully compatible. The new one breaks tests for `var()`.

Can you show the failures you get? Indeed it would be nice if we were able to use a single implementation.

Reply:
>> Is that promotion exactly identical in all cases to _mean_promote()?

> Well the situation there is quite messy, and there's a very old comment saying that things should be improved. 

Right now seems like a good opportunity to do so.

> Can you show the failures you get? 

You know, funny thing happened, I tried to reproduce the failures and all tests passed without complaint. I pushed the implementation not reliant on  ```Base.reducedim_init()```, take a look.